File: solo/puzzles/easy/brackets-extreme-edition/BracketsExtremeEdition.cpp
#include <iostream>
#include <string>
#include <stack>
#include <map>

int main()
{
   std::string expression;
   std::cin >> expression; std::cin.ignore();

   std::stack<char> stack;
   std::map<char, char> pairs = { { '{', '}' }, { '(', ')' }, { '[', ']' } }; 

   for (char bracket : expression)
   {
      if (pairs.find(bracket) != pairs.end())
      {
         stack.push(bracket);
      }
      else if (bracket == '}' || bracket == ')' || bracket == ']')
      {
         if (stack.empty())
         {
            std::cout << "false" << std::endl;
            return 0;
         }

         char open_bracket = stack.top();
         stack.pop();
         std::cerr << "open: " << open_bracket << " close: " << bracket << std::endl;
         if (bracket != pairs[open_bracket])
         {
            std::cout << "false" << std::endl;
            return 0;
         }
      }
   }

   if (stack.empty())
   {
      std::cout << "true" << std::endl;
   }
   else
   {
      std::cout << "false" << std::endl;
   }
   return 0;
}

File: solo/puzzles/easy/brackets-extreme-edition/brackets_extreme_edition.rs
use std::io;
use std::collections::HashMap;

macro_rules! parse_input {
   ($x:expr, $t:ident) => ($x.trim().parse::<$t>().unwrap())
}

fn main() {
   let mut input_line = String::new();
   io::stdin().read_line(&mut input_line).unwrap();
   let expression = input_line.trim().to_string();
   let mut stack = Vec::new();
   let pairs: HashMap<char, char> = [('(', ')'), ('[', ']'), ('{', '}')].iter().cloned().collect();

   for bracket in expression.chars() {
      if pairs.contains_key(&bracket) {
         stack.push(bracket);
      } else if bracket == ')' || bracket == ']' || bracket == '}' {
         if stack.is_empty() {
            println!("false");
            return;
         }

         let open_bracket = stack.pop().unwrap();
         if pairs[&open_bracket] != bracket {
            println!("false");
            return;
         }
      }
   }

   if stack.is_empty() {
      println!("true");
   } else {
      println!("false");
   }
}
